versioned_docs/version-4.0.0-alpha.1/datastructures/storage-in-metadata.md

```diff
@@ -3,6 +3,89 @@ title: Metadata Format
 slug: /datastructures/storage-in-metadata
 ---
 
-:::caution
-TODO: Add documentation on the `storage ` field in the metadata.
-:::
+The storage layout of a contract is reflected inside the metadata. It allows third-party 
+tooling to work with contract stroage and can also help to better understand the storage 
+layout of any given contract.
+
+Given a contract with the following storage:
+
+```rust
+#[ink(storage)]
+pub struct MyContract {
+    balance: Balance,
+    block: BlockNumber,
+    something: Lazy<bool>,
+}
+```
+
+The storage will be reflected inside the metadata as like follows:
+
+```json
+"root": {
+  "layout": {
+    "struct": {
+      "fields": [
+        {
+          "layout": {
+            "leaf": {
+              "key": "0x00000000",
+              "ty": 0
+            }
+          },
+          "name": "balance"
+        },
+        {
+          "layout": {
+            "leaf": {
+              "key": "0x00000000",
+              "ty": 1
+            }
+          },
+          "name": "block"
+        },
+        {
+          "layout": {
+            "root": {
+              "layout": {
+                "leaf": {
+                  "key": "0xb1f4904e",
+                  "ty": 2
+                }
+              },
+              "root_key": "0xb1f4904e"
+            }
+          },
+          "name": "something"
+        }
+      ],
+      "name": "MyContract"
+    }
+  },
+  "root_key": "0x00000000"
+}
+```
+
+We observe that the storage is layed out as a tree, where tangible storage values end up 
+inside a `leaf`. Because of `Packed` encoding, leafs can share the same storage key, and 
+in order to reach them you'd need fetch and decode the whole storage cell under this key.
+
+A `root_key` is meant to either be used to directly access a `Packed` storage field 
+or to serve as the base key for calculting the actual keys needed to access values in 
+`Non-Packed` fields (such as Mappings).
+
+## Storage key calculation for mappings
+
+Root storage keys are always 4 bytes in size. However, the storage API of the contracts 
+pallet supports keys of arbitrary length. In order to reach a mapping value, the storage 
+key of said value is calculated at runtime.
+
+The formula to calculate a storage key `S` used to access a mapping value under the key `K`
+for a mapping with base key `B` can be expressed as follows:
+
+```
+S = B + scale::encode(K)
+```
+ 
+In words, find the base (root) key of the mapping and concatenate it with the 
+SCALE encoded key of the mapped value to obtain the actual storage key used to 
+access the mapped value.
```

Comment:
Isn't there a hashing done as well?

Reply:
Oh, good catch! Yes, IIRC in the latest versions of pallet contracts the key is blake2 hashed and append to its own hash to that the key again. You are very right, if you were try to access storage off-chain (without going through the pallet storage API) , I should absolutely documents that here!

Reply:
<img width="650" alt="image" src="https://user-images.githubusercontent.com/18346821/215488461-c83c0ad7-7338-481e-b024-6f7d1f3bc98a.png">

I want to highlight that a new version uses `Blake2_128Concat`. I don't know whether we need to mention it in the documentation or not.

Reply:
I was also asking myself if it does really fit into the docs. But I came to the conclusion that it should be documented, as the storage layout is part of the contract metadata. I already just wrote a few paragraphs about it. Figuring out how this exactly works requires reading through a lot of substrate code otherwise.